optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintFactory.java

```diff
@@ -63,7 +68,15 @@ public DroolsConstraintFactory(SolutionDescriptor<Solution_> solutionDescriptor)
 
     @Override
     public <A> BiConstraintStream<A, A> fromUniquePair(Class<A> fromClass, BiJoiner<A, A> joiner) {
-        throw new UnsupportedOperationException();
+        MemberAccessor planningIdMemberAccessor = ConfigUtils.findPlanningIdMemberAccessor(fromClass);
+        if (planningIdMemberAccessor == null) {
+            throw new IllegalArgumentException("The fromClass (" + fromClass + ") has no member with a @"
+                    + PlanningId.class.getSimpleName() + " annotation,"
+                    + " so the pairs can be made unique ([A,B] vs [B,A]).");
+        }
+        // TODO Breaks node sharing + involves unneeded indirection
```

Comment:
The TODO likely doesn't apply. I copy-pasted this code from Bavet impl.

Reply:
Sounds like a good unit test - which drools could win while bavet wouldn't.

Ignored tests can show how the impl differ in completeness :)

Reply:
Actually, I don't think we can test for this in Drools. What Drools does to the RETE network is an implementation detail that we don't really want to deal with. Personally, I don't think we should worry about that - we should write clean good rules, and the rest is Drools internal business.